connector/src/main/scala/com/datastax/spark/connector/writer/AsyncExecutor.scala

```diff
@@ -97,8 +95,10 @@ class AsyncExecutor[T, R](asyncAction: T => CompletionStage[R], maxConcurrentTas
     * It will not wait for tasks scheduled for execution during this method call,
     * nor tasks for which the [[executeAsync]] method did not complete. */
   def waitForCurrentlyExecutingTasks() {
-    for ((future, _) <- pendingFutures.snapshot())
+    while(!pendingFutures.isEmpty){
+      val future = pendingFutures.poll()
       Try(Await.result(future, Duration.Inf))
+    }
```

Comment:
_Deleted my first comment where I stupidly misread the while loop… however:_

We have changed the semantic here according to the documentation: previously it was a snapshot so it would deterministically wait for only the _currently_ executing tasks as of the time the method is invoked. Now `pendingFutures` can continue to admit new futures in the queue while we are looping, so it's non-deterministic, the outer method could theoretically never terminate.